kong/tracing/propagation/injectors/_base.lua

```diff
@@ -0,0 +1,213 @@
+      return true
+    end
+  end
+
+  return false, "no required field found in context: " ..
+                table.concat(any, ", ")
+end
+
+
+function _INJECTOR:verify_all(context)
+  local all = self.context_validate.all
+  if not all or #all == 0 then
+    return true
+  end
+
+  if not context or type(context) ~= "table" then
+    return false, "no context to inject"
+  end
+
+  for _, field in ipairs(all) do
+    if context[field] == nil then
+      return false, "field " .. field .. " not found in context"
+    end
+  end
+
+  return true
+end
+
+
+-- injection failures are reported, injectors are not expected to fail because
+-- kong should ensure the tracing context is valid
+function _INJECTOR:verify_context(context)
+  local ok_any, err_any = self:verify_any(context)
+  local ok_all, err_all = self:verify_all(context)
+
+  if ok_any and ok_all then
+    return true
+  end
+
+  local err = err_any or ""
+  if err_all then
+    err = err .. (err_any and ", " or "") .. err_all
+  end
+
+  return false, err
+end
+
+
+function _INJECTOR:inject(inj_tracing_ctx)
+  local context_verified, err = self:verify_context(inj_tracing_ctx)
+  if not context_verified then
+    return nil, self.name ..  " injector context is invalid: " .. err
+  end
+
+  -- Convert IDs to be compatible to the injector's format.
+  -- Use trace_id_allowed_sizes to try to keep the original (incoming) size
+  -- where possible.
+  -- Extractors automatically set `trace_id_original_size` during extraction.
+  local orig_size = inj_tracing_ctx.trace_id_original_size
+  local allowed = self.trace_id_allowed_sizes
+  local new_trace_id_size = contains(allowed, orig_size) and orig_size
+      or allowed[1]
+
+  inj_tracing_ctx.trace_id  = to_id_size(inj_tracing_ctx.trace_id, new_trace_id_size)
+  inj_tracing_ctx.span_id   = to_id_size(inj_tracing_ctx.span_id, self.span_id_size_bytes)
+  inj_tracing_ctx.parent_id = to_id_size(inj_tracing_ctx.parent_id, self.span_id_size_bytes)
+
+  local headers, h_err = self:create_headers(inj_tracing_ctx)
+  if not headers then
+    return nil, h_err
+  end
+
+  for _, header in ipairs(headers) do
+    set_header(header.name, header.value)
+  end
+
+  local formatted_trace_id, t_err = self:get_formatted_trace_id(inj_tracing_ctx.trace_id)
+  if not formatted_trace_id then
+    return nil, t_err
+  end
+  return formatted_trace_id
+end
+
+
+--- Create headers to be injected.
+--
+-- Function to be implemented by Injector subclasses, uses the extracted
+-- tracing context to create and return headers for injection.
+--
+-- @function _INJECTOR:create_headers(tracing_ctx)
+-- @param table tracing_ctx The extracted tracing context.
+--   The structure of this table is described in the Extractor base class.
+-- @return table/array-of-tables that define the headers to be injected
+--   example:
+--   return {
+--     {
+--       name = "traceparent",
+--       value = "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01"
+--     }
+--   }
```

Comment:
Why normal dictionary does not work here? Like

```lua
{ ["traceparent"]="00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01" }
```

Reply:
an older implementation had more information than just name and value, it can be simplified now as you say, thanks.